lib/controllers/book_controller.dart

```diff
@@ -8,49 +8,123 @@ class LessonState {
   final bool loading;
 
   final List<BookModel> books;
+  final List<String> subjects;
+  final List<String> levels;
 
   const LessonState({
     required this.loading,
     required this.books,
+    required this.subjects,
+    required this.levels,
   });
 
   LessonState copyWith({
     bool? loading,
     List<BookModel>? books,
+    List<String>? subjects,
+    List<String>? levels,
   }) {
     return LessonState(
       loading: loading ?? this.loading,
       books: books ?? this.books,
+      subjects: subjects ?? this.subjects,
+      levels: levels ?? this.levels,
     );
   }
 }
 
 class LessonController extends StateNotifier<LessonState> {
-  LessonController(this._lessonUseCase) : super(const LessonState(loading: false, books: []));
+  LessonController(this._lessonUseCase) : super(const LessonState(loading: false, books: [], subjects: [], levels: []));
   final LessonUseCase _lessonUseCase;
 
   Future<void> fetchInitialBooks() async {
-    print("fetch initial books");
-
     state = state.copyWith(loading: true);
 
     List<BookModel> bookResult = await _lessonUseCase.initBooks();
+
+    bookResult.removeWhere((book) => book.displayTitle == null);
+    List<String> subjects = [];
     for (var book in bookResult) {
-      print(book.url);
+      for (var subject in book.subjects) {
+        if (!subjects.contains(subject)) {
+          subjects.add(subject);
+        }
+      }
     }
-    bookResult.removeWhere((book) => book.displayTitle == null);
+    List<String> levels = [];
+    for (var book in bookResult) {
+      for (var level in book.levels) {
+        if (!levels.contains(level)) {
+          levels.add(level);
+        }
+      }
+    }
+
+    sort(bookResult);
 
     state = state.copyWith(
       loading: false,
       books: bookResult,
+      subjects: subjects,
+      levels: levels,
     );
   }
 
+  void sort(List<BookModel> bookResult) {
+    sortBySubject(bookResult);
+    sortByLevel(bookResult);
+  }
+
+  void sortByLevel(List<BookModel> bookResult) {
+    bookResult.sort((a, b) {
+      String levelA = a.levels.isNotEmpty ? a.levels[0] : '';
+      String levelB = b.levels.isNotEmpty ? b.levels[0] : '';
+      return levelA.compareTo(levelB);
+    });
+  }
+
+  void sortBySubject(List<BookModel> bookResult) {
+    bookResult.sort((a, b) {
+      String subjectA = a.subjects.isNotEmpty ? a.subjects[0] : '';
+      String subjectB = b.subjects.isNotEmpty ? b.subjects[0] : '';
+      return subjectA.compareTo(subjectB);
+    });
+  }
+
   Future<void>? resetAll() {
     state = state.copyWith(
       loading: false,
     );
 
     return null;
   }
+
+  Future<void> fetchBooksByFilter({String? subject, String? level}) async {
+    state = state.copyWith(loading: true);
+
+    List<BookModel> filteredBooks = [];
+    List<BookModel> bookResult = await _lessonUseCase.initBooks();
```

Comment:
Re-fetching all the books on each change of either of the dropdown menus seems wasteful, when the complete list of books can be fetched once, and a filtered view be computed in case of a change.